### DISCUSSION TITLE
How do you exclude source files during unit tests?

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1560] [Q&A]
I have a component A whose source is added to `SOURCE_FILES` and a number of other sources which come from an external project, and which are only used in the current work in the context of component A. They are also added to `SOURCE_FILES` in component A's CMakeLists.txt.
```cmake
list(APPEND SOURCE_FILES
    "path/to/external/source1.c"
    "path/to/external/source2.c"
    "path/to/external/source3.c"
```

The functions contained in some of those sources need to be mocked in the unit tests, so:
```cmake
set(UT_SOURCE_FILES
    "path/to/external/source1_mocks.c"
    "path/to/external/source2_mocks.c"
    "path/to/external/source3_mocks.c"
    "${CMAKE_CURRENT_LIST_DIR}/tests/ut/TestMain.cpp"
    "${CMAKE_CURRENT_LIST_DIR}/tests/ut/Tester.cpp"
    "${CMAKE_CURRENT_LIST_DIR}/tests/ut/TesterBase.cpp"
    "${CMAKE_CURRENT_LIST_DIR}/tests/ut/GTestBase.cpp"
)
```

But, in this circumstance, I get "multiple definition of " errors for all the functions shared between `sourcex.c` and `sourcex_mocks.c`.

How can I add `sourcex.c` to the standard build target, but not the unit-test target?

**Additional**
It seems to suggest in the commentary on `register_fprime_ut` in `API.cmake` that `SOURCE_FILES` is not added to `UT_SOURCE_FILES`, because the comment specifically asks for the `*Ai.xml` file and implementation source, but, at the bottom of the function, there is this line:
`setup_all_module_targets(FPRIME_UT_TARGET_LIST ${MODULE_NAME} "" "${SOURCE_FILES}" "${AC_OUTPUTS}" "${MD_IFS}")`

Is this a mistake? Should `UT_SOURCE_FILES` including `SOURCE_FILES`?

**Comment by timcanham** [comment]
If you've included `sourcex.c` in your compilation of the component, then it's considered code that needs to be tested with the component. If you want to provide test versions of the `sourcex.c` files, I would put them in a different directory as a library used by the component. Then you can substitute your own test code.
